scalanet/discovery/src/io/iohk/scalanet/discovery/ethereum/v4/DiscoveryRPC.scala

```diff
@@ -0,0 +1,58 @@
+package io.iohk.scalanet.discovery.ethereum.v4
+
+import io.iohk.scalanet.discovery.crypto.{PublicKey}
+import io.iohk.scalanet.discovery.ethereum.{Node, EthereumNodeRecord}
+import monix.eval.Task
+
+/** The RPC method comprising the Discovery protocol between peers. */
+trait DiscoveryRPC[A] {
+  import DiscoveryRPC.{Call, Proc}
+
+  /** Sends a Ping request to the node, waits for the correct Pong response,
+    * and returns the ENR sequence, if the Pong had one.
+    */
+  def ping: Call[A, Proc.Ping]
+
+  /** Sends a FindNode request to the node and collects Neighbours responses
+    * until a timeout or if the maximum expected number of nodes are returned.
+    */
+  def findNode: Call[A, Proc.FindNode]
+
+  /** Sends an ENRRequest to the node and waits for the correct ENRResponse,
+    * returning the ENR from it.
+    */
+  def enrRequest: Call[A, Proc.ENRRequest]
+}
+
+object DiscoveryRPC {
+  type ENRSeq = Long
+
+  /** Pair up requests with responses in the RPC. */
+  sealed trait Proc {
+    type Req
+    type Res
+  }
+  object Proc {
+    trait Ping extends Proc {
```

Comment:
I find it a little not intuitive that we alreagy have  already concrete `Ping` case class and we need to redifine it here as trait

Reply:
I didn't want to come up with a new name, I only use it as `Proc.Ping` to distinguish, I thought it may be better than `PingProc` but I might be wrong. I wanted to capture the symmetric relationship between the remote and local RPC interfaces and the `Call` semantics without having to repeat the call signatures everywhere.